### PR TITLE
(fixes #157) uses a separate workflow/dependency graph to manage dependency building

### LINKED ISSUE #157: Re-examine how we do dependencies
The fix of https://github.com/krux/hyperion/pull/155 introduces in consistent behavior to pipeline activities where most other methods returns a copy of the object but dependsOn does not. Examine and act on making it consistent again. Potential solutions:

- Decouple dependsOn from activity objects and make it part of workflow expression.
- Workflow performs a "lazy build", and only performs one dependsOn per object.

Also the way we fetch all involved objects in workflow now becomes outdated, as we are not returning the last object that have all the dependencies in workflow, but all activities. For example, there is no need to return dependsOn in `def objects`.

### PR DESCRIPTION
**Pull request by realstraw**
(fixes #157) uses a separate workflow/dependency graph to manage dependency building

### CODE REVIEW

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import com.krux.hyperion.activity.PipelineActivity
+import com.krux.hyperion.common.PipelineObjectId
+import scala.collection.mutable.Buffer
+import scala.language.implicitConversions
+import scala.annotation.tailrec
+
+class WorkflowGraph private (
+  val flow: Map[PipelineObjectId, Set[PipelineObjectId]],
+  val activities: Map[PipelineObjectId, PipelineActivity]
+  // val roots: Set[PipelineObjectId]
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Remove dead code.

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+) {
+
+  def this() =
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Why make the constructor above private and then have this?  Is this better than assigning default values to the constructor above?  Perhaps it is - just a Q.

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+   * Merges two workflow graphs
+   */
+  def ++(other: WorkflowGraph): WorkflowGraph = {
+    val newFlow = mergeFlow(this.flow, other.flow)
+    val newActivities = this.activities ++ other.activities
+    new WorkflowGraph(newFlow, newActivities)
+  }
+
+  private def mergeFlow(
+    flow1: Map[PipelineObjectId, Set[PipelineObjectId]],
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Can't you define a `type` above for this Map type to clean up the code a bit?

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala (hunk 1 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Oops, will do.

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  final def toActivities: Iterable[PipelineActivity] = {
+    assert(roots.size != 0)
+    assert(leaves.size != 0)
+
+    if (flow.size == 0) {
+      activities.values
+    } else {
+      // get the immediate dependencies from the root node
+      val rootDependents: Set[(PipelineObjectId, PipelineObjectId)] =
+        for { act <- (roots -- isolates); dependent <- flow(act) } yield (dependent, act)
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Nit: would be clearer for the iterators to be on a separate line, like so:
```
for {
  act <- (roots -- isolates)
  dependent <- flow(act)
} yield (dependent, act)
```

**Comment by sethyates** [comment]
:+1: although I can't vouch for the correctness of the `WorkflowGraph` code just from a static inspection, so assuming you're confident the unit tests cover the core and edge cases, then LGTM :sparkles: :sparkles: :sparkles:

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala (hunk 2 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Because I don't want other classes that uses workflow graph to set `flow` or `activities`, they are managed by `+`, `++`, and `~>`

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala (hunk 3 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Good idea, will do

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala (hunk 4 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: OK, will do

File: core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala (hunk 2 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/util/WorkflowGraph.scala]: Ah, strewth.  Duh.
